app/src/main/java/clarkson/ee408/tictactoev4/MainActivity.java

```diff
@@ -26,20 +31,89 @@ public class MainActivity extends AppCompatActivity {
     private Button[][] buttons;
     private TextView status;
     private Gson gson;
+    private boolean shouldRequestMove;
+    private SocketClient socketClient;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         this.tttGame = new TicTacToe(STARTING_PLAYER_NUMBER);
         this.gson = new GsonBuilder().serializeNulls().create();
+        socketClient = SocketClient.getInstance();
+        shouldRequestMove = false;
 
         buildGuiByCode();
+        updateTurnStatus();
 
         Handler handler = new Handler();
         GameMoveTaskRunnable runnable = new GameMoveTaskRunnable(this, handler);
         handler.post(runnable);
     }
 
+    /**
++   * Sends a request to the server to ask for a game move made by the other player.
++   */
+    public void requestMove() {
+        if (!shouldRequestMove) {
+            return; // Only request moves when it's our turn
+        }
+
+        // Create Request object with type REQUEST_MOVE
+        Request request = new Request();
+        request.setType(RequestType.REQUEST_MOVE);
+        // You might want to include game state or player info
+        request.setData(""); // Add any necessary data
+
+        // Use SocketClient to send request in networkIO thread
+        AppExecutors.getInstance().networkIO().execute(() -> {
+            try {
+                GamingResponse response = socketClient.sendRequest(request, GamingResponse.class);
+
+                // Process response in main thread
+                AppExecutors.getInstance().mainThread().execute(() -> {
+                    if (response != null && response.getStatus() == ResponseStatus.SUCCESS) {
+                        // Get the move from GamingResponse (already parsed)
+                        int moveValue = response.getMove();
+
+                        // Validate move value
+                        if (moveValue >= 0 && moveValue < TicTacToe.SIDE * TicTacToe.SIDE) {
+                            // Convert single integer move to row and column
+                            int row = moveValue / TicTacToe.SIDE;
+                            int col = moveValue % TicTacToe.SIDE;
+
+                            // Check if the position is available
+                            if (tttGame.getBoard()[row][col] == 0) {
+                                // Utilize update() function to add changes to the board
+                                update(row, col);
+                            }
```

Comment:
There is no `getBoard()` method in tttGame. Just call update with the row and col.

```suggestion
                            // Utilize update() function to add changes to the board
                            update(row, col);
```